Add strong_migrations gem

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [X] Stop us from shooting ourselves in the foot

## Description

I've had good experiences with the [strong_migrations](https://github.com/ankane/strong_migrations) gem in the past. From the description: 

> Catch unsafe migrations in development 

This is how a potential warning looks like:

![screenshot of warning](https://camo.githubusercontent.com/fca6bf62c62bf157b040d9c4d1ceb89853e5a721/68747470733a2f2f616e6b616e652e6f72672f696d616765732f7374726f6e672d6d6967726174696f6e732e706e67)

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [X] no documentation needed

## Additional context

I recommend reading [PostgreSQL rocks, except when it blocks: Understanding locks](https://www.citusdata.com/blog/2018/02/15/when-postgresql-blocks/) for more information on some potentially problematic migrations.

rhymes commented:
I like this gem, especially now that we've migrated to PostgreSQL 11 (hopefully to 12 at some point), does it only emit warning or does it block the migration?

I wonder if it would break `rails db:reset` or the setup of new contributors because we probably have some older migrations which might be blocked by this gem, though I see there's a way to set it after the last migration we have in the code: https://github.com/ankane/strong_migrations#existing-migrations

I would also add this check to make sure we only run checks targeted to the correct PostgreSQL version: https://github.com/ankane/strong_migrations#target-version

Let's add those at least :)

mstruve commented:
Probably want to implement some of @rhymes suggestions otherwise I love this idea! 

Question, is there a way to run this in test as well? My guess is you will get some people that might not run migrations locally first and it would be nice if CircleCi caught it as well. What do you think?

citizen428 commented:
@rhymes Both valid points, added in `config/initializers/strong_migrations.rb`, thanks 😃 

@maestromac Thanks for confirming.

@mstruve I moved the gem to the main group, as suggested in their README. Based on the description of their README, it should definitely run during tests:

> For safety, this option only affects development and test environments. In other environments, the actual server version is always used.

Some other things I'd like to discuss/point out:

* Should we add [schema sanity](https://github.com/ankane/strong_migrations#schema-sanity)? It'd be good for my OCD for sure, it drives me mad when schema files randomly change. 😅 
* Same for [analyze tables](https://github.com/ankane/strong_migrations#analyze-tables), yay/nay/later?
* Just a heads up to @mstruve that dangerous tasks are disabled in production. In case you ever hit this and need to override it, [here is how](https://github.com/ankane/strong_migrations#dangerous-tasks).

citizen428 commented on config/initializers/strong_migrations.rb:
```diff
@@ -0,0 +1,5 @@
+# https://github.com/ankane/strong_migrations#existing-migrations
+StrongMigrations.start_after = 2019_12_27_114543
```
While it may look unusual, this is the way the timestamp gets formatted in `db/schema.rb`, so I thought it's easiest to keep it the same way.

mstruve commented:
> Should we add schema sanity? 
I am totally game for this.

>Same for analyze tables?
Nay 

I'm also with @rhymes let's keep it out of production and add it to the `group :development, :test do` block

citizen428 commented:
> I'm not convinced we need to have it in the main group and having it load in production. 

It intercepts potentially dangerous operations from happening in production, which you need to manually override to signal that you know what you're doing. But point taken, I moved it to `group :development, :test`.

No other changes for now, let's add schema sanity once we're all on board with keeping the gem around.
